### PR TITLE
feat: directory checkpoint storage [DET-9594]

### PR DESCRIPTION
**Pull request by ioga**
## Description

Introduce a new checkpoint storage type `directory` which writes the checkpoints and tensorboard files to a local directory.

Intended use cases are 1) using k8s PVC for ckpt storage 2) better representing detached mode storage paradigm

See included docs for the usage details.

## Test Plan

Automated test for bind-mounted storage on agent RM is included:
1. bind mount a directory into the experiment
2. configure the same directory for storage using the new option
3. make sure the data is saved to the mounted directory

k8s pod spec setup:
1. make a PVC inside a k8s cluster, e.g. on GKE.
2. mount it into an experiment using the pod_spec.
3. make sure the data is on the PVC

For example, using a GKE cluster, make a yaml for PVC and create it using `kubectl create -f <filename>`
```
# pvc-demo.yaml
apiVersion: v1
kind: PersistentVolumeClaim
metadata:
  name: pvc-demo
spec:
  accessModes:
    - ReadWriteOnce
  resources:
    requests:
      storage: 10Gi
  storageClassName: standard-rwo
```

Run an experiment with the following extra content in the config, which'll mount and use the PVC. Since our PVC is a simple GKE PVC, it can only be mounted on a single node, so this must be a single slot experiment (or multi-slot but single-node, if you're brave enough). 
```
# appended to your exp config yaml
checkpoint_storage:
  type: directory
  container_path: /tmp/somepath2/

environment:
  pod_spec:
    apiVersion: v1
    kind: Pod
    spec:
      volumes:
        - name: pvc-demo-vol
          persistentVolumeClaim:
            claimName: pvc-demo
      containers:
        - name: determined-container
          volumeMounts:
            - name: pvc-demo-vol
              mountPath: /tmp/somepath2
```

Then, run a notebook using the same PVC to check it's there:
```
# pvc-notebook.yaml
environment:
  pod_spec:
    apiVersion: v1
    kind: Pod
    spec:
      volumes:
        - name: pvc-demo-vol
          persistentVolumeClaim:
            claimName: pvc-demo
      containers:
        - name: determined-container
          volumeMounts:
            - name: pvc-demo-vol
              mountPath: /tmp/somepath2
```

`det notebook start --config resoureces.slots=0 --config-file pvc-notebook.yaml`

In the notebook in `/tmp/somepath2`, you should see a bunch of uuids for the checkpoints.
Inside the notebook, you can run `det e list-checkpoints <experiment id>`, find a checkpoint uuid, and then `det checkpoint download <checkpoint uuid>`. this should "download" the checkpoint into the current directory.

## Checklist

- [ ] Changes have been manually QA'd
- [ ] User-facing API changes need the "User-facing API Change" label.
- [ ] Release notes should be added as a separate file under `docs/release-notes/`.
  See [Release Note](https://github.com/determined-ai/determined/blob/master/docs/release-notes/README.md) for details.
- [ ] Licenses should be included for new code which was copied and/or modified from any external code.

### CODE REVIEW

**Review by stoksc — APPROVED** [review]
backend +1

File: e2e_tests/tests/command/test_tensorboard.py [hunk 1]
```diff
@@ -154,7 +156,7 @@ def test_start_tensorboard_for_multi_experiment(tmp_path: Path, secrets: Dict[st
 
 
 @pytest.mark.e2e_cpu
-def test_start_tensorboard_with_custom_image(tmp_path: Path) -> None:
+def test_start_tensorboard_with_custom_image() -> None:
```
Comment (mpkouznetsov) [inline e2e_tests/tests/command/test_tensorboard.py:159]: 👍

**Review by mpkouznetsov — APPROVED** [review]
LGTM

File: docs/reference/deploy/helm-config-reference.rst [hunk 2]
```diff
@@ -128,6 +128,13 @@
       -  ``hostPath``: The file system path on each node to use. This directory will be mounted to
          ``/determined_shared_fs`` inside the trial pod.
 
+   -  ``type: directory``: Checkpoints are written to a local directory in the task container. Users
+      are responsible for mounting a persistent storage at this path, e.g., a shared PVC using
+      ``pod_spec`` configuration.
+
+      -  ``containerPath``: The file system path inside the task pods to use. The checkpoints and
+         tensorboard data will be stored under this path.
+
```
Comment (tara-det-ai) [inline docs/reference/deploy/helm-config-reference.rst:137]: tensorboard should be TensorBoard in all instances in the text

File: docs/reference/deploy/master-config-reference.rst [hunk 3]
```diff
@@ -1286,6 +1286,35 @@ When an experiment finishes, the system will optionally delete some checkpoints
...
+   experiments. Therefore, if an experiment is using ``type: directory`` storage, and mounts the
+   storage separately, a launched TensorBoard will need the same mount configuration provided
+   explicitly using ``det tensorboard start <experiment_id> --config-file <CONFIG FILE>`` or
+   similar.
+
+.. warning::
+
+   When downloading checkpoints (e.g., using ``det checkpoint download``), we assume the same
+   directory is present locally at the same ``container_path``.
+
```
Comment (tara-det-ai) [inline docs/reference/deploy/master-config-reference.rst:1312]: .. warning::

   When downloading checkpoints (e.g., using ``det checkpoint download``), Determined assumes the same
   directory is present locally at the same ``container_path``.

File: docs/reference/training/experiment-config-reference.rst [hunk 4]
```diff
@@ -619,6 +619,36 @@ Optional. `Propagation behavior
...
+   experiments. Therefore, if an experiment is using ``type: directory`` storage, and mounts the
+   storage separately, a launched TensorBoard will need the same mount configuration provided
+   explicitly using ``det tensorboard start <experiment_id> --config-file <CONFIG FILE>`` or
+   similar.
+
+.. warning::
+
+   When downloading checkpoints (e.g., using ``det checkpoint download``), we assume the same
+   directory is present locally at the same ``container_path``.
+
```
Comment (tara-det-ai) [inline docs/reference/training/experiment-config-reference.rst:646]: .. warning::

   When downloading checkpoints (e.g., using ``det checkpoint download``), Determined assumes the same
   directory is present locally at the same ``container_path``.

**Review by tara-det-ai — COMMENTED** [review]
suggested edits

**Comment by ioga** [comment]
merging over a lint-react flake

**Pull request merged**